neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,68 +1,160 @@
 from typing import Callable, List, Dict, Optional, Set, cast
 
+import scipy
 import numpy as np
 from typeguard import check_argument_types
 
 from neuralmonkey.model.model_part import ModelPart
-from neuralmonkey.decoders.beam_search_decoder import (BeamSearchDecoder,
-                                                       SearchStepOutput)
+from neuralmonkey.decoders.beam_search_decoder import BeamSearchDecoder
 from neuralmonkey.runners.base_runner import (BaseRunner, Executable,
                                               ExecutionResult, NextExecute)
-from neuralmonkey.vocabulary import Vocabulary, END_TOKEN
+# pylint: disable=unused-import
+from neuralmonkey.runners.base_runner import FeedDict
+# pylint: enable=unused-import
+from neuralmonkey.vocabulary import PAD_TOKEN_INDEX, END_TOKEN, PAD_TOKEN
 
 
 class BeamSearchExecutable(Executable):
     def __init__(self,
                  rank: int,
                  all_encoders: Set[ModelPart],
-                 bs_outputs: SearchStepOutput,
-                 vocabulary: Vocabulary,
+                 num_sessions: int,
+                 decoder: BeamSearchDecoder,
                  postprocess: Optional[Callable]) -> None:
 
         self._rank = rank
+        self._num_sessions = num_sessions
         self._all_encoders = all_encoders
-        self._bs_outputs = bs_outputs
-        self._vocabulary = vocabulary
+        self._decoder = decoder
         self._postprocess = postprocess
 
-        self.result = None  # type: ExecutionResult
+        self.step = 0
+        self.scores = np.empty([0, decoder.beam_size], dtype=float)
+        self.parent_ids = np.empty([0, decoder.beam_size], dtype=int)
+        self.token_ids = np.empty([0, decoder.beam_size], dtype=int)
 
-    def next_to_execute(self) -> NextExecute:
-        return self._all_encoders, {"bs_outputs": self._bs_outputs}, {}
+        self._next_feed = [{} for _ in range(self._num_sessions)] \
+            # type: List[FeedDict]
 
-    def collect_results(self, results: List[Dict]) -> None:
-        if len(results) > 1:
-            raise ValueError("Beam search runner does not support ensembling.")
+        # If we are ensembling, we run only one step at a time
+        # Note, that in the first iteration, we only want to generate
+        # logprobs for ensembling, therefore we set max_steps = 0
```

Comment:
moc tomuhle komentari nerozumim - proč max steps 0 a ne 1?

Reply:
Jeste to zkusim preformulovat. Je to dusledek toho, jakym zpusobem je implementovany beamsearch_decoder.py